pennylane_ls/FermionDevice.py

```diff
@@ -64,67 +75,94 @@ def apply(self, operation, wires, par):
+        mean = np.mean(shots, axis=0)
+        return sum(mean[wires.tolist()])
 
     def var(self, observable=None, wires=None, par=None):
         """
         Retrieve the requested observable variance.
         """
-        try:
-            shots = self.sample(observable, wires, par)
-            return np.var(shots, axis=0)
-        except:
-            raise NotImplementedError()
+        if self._observable_map[observable] == Identity:
+            return 0.0
+
+        shots = self.sample(observable, wires, par)
+        if self._observable_map[observable] == PauliZ:
+            shots = np.ones(shots.shape) - 2*shots
+        var = np.var(shots, axis=0)
+        return sum(var[wires.tolist()])
 
     def sample(self, observable, wires, par):
         """
         Retrieve the requested observable expectation value.
         """
         observable_class = self._observable_map[observable]
         if issubclass(observable_class, FermionObservable):
-            # submit the job
-            wires = wires.tolist()
-            for wire in wires:
-                m_obj = ('measure', [wire], [])
-                self.job_payload['experiment_0']['instructions'].append(m_obj)
-            url= self.url_prefix + "post_job/"
-            job_response = requests.post(url, data={'json':json.dumps(self.job_payload),
-                                                         'username': self.username,'password':self.password})
-
-            job_id = (job_response.json())['job_id']
-
-            # obtain the job result
-            result_payload = {'job_id': job_id}
-            url= self.url_prefix + "get_job_result/"
-
-            result_response = requests.get(url, params={'json':json.dumps(result_payload),
-                                                        'username': self.username,'password':self.password})
-            results_dict = json.loads(result_response.text)
-            results = results_dict["results"][0]['data']['memory']
-
-            num_obs = len(wires)
-            out = np.zeros((self.shots,num_obs))
-            for i1 in np.arange(self.shots):
-                temp = results[i1].split()
-                for i2 in np.arange(num_obs):
-                    out[i1,i2] = int(temp[i2])
-            return out
+            return self._samples
         raise NotImplementedError()
 
+    def probability(self, wires=None):
+        shots = self._samples
+        if wires is not None:
+            shots = shots[:,wires.tolist()]
+
+        #np.sort(np.unique(shots, axis=0, return_counts=True, dtype=[('pattern', ), ('probability', )]), order='pattern')
+
+        patterns, probabilities = np.unique(shots, axis=0, return_counts=True)
+        sort_labels = np.argsort(patterns, axis=0)
+        patterns = patterns[sort_labels]
+        probabilities = probabilities[sort_labels]
+
+        return OrderedDict(zip(map(tuple, patterns), probabilities))
+
+    def pre_measure(self):
```

Comment:
no idea what is going on here. I assume that @TheNewYorkerFlute has some strong opinions.

Reply:
This is just a conversion to the right data format the PennyLane wants. With this the pennylane.probs() return type is supported by the device and yields correct probabilities.

Reply:
@cvjjm, I think we can get probs nicely by converting the unique patterns into decimal values and then sorting the decimal values array. I tested it and this nicely gives an ordered dict. The code is:

patterns, probabilities = np.unique(arr, axis=0, return_counts=True)

patterns_decimal_repr=np.packbits(patterns.astype('int32'), axis=1)
patterns_decimal_repr=patterns_decimal_repr.ravel()

sort_labels = np.argsort(patterns_decimal_repr, axis=0)

patterns = patterns[sort_labels]
probabilities = probabilities[sort_labels]

odict=OrderedDict(zip(map(tuple, patterns), probabilities))

Reply:
And also:
probabilities = probabilities/probabilities.sum()

Reply:
This is a nice solution!